plugins/woocommerce/src/Admin/API/Reports/GenericStatsController.php

```diff
@@ -155,4 +165,60 @@ public function get_item_schema() {
 			),
 		);
 	}
+
+	/**
+	 * Get all reports.
+	 *
+	 * @param \WP_REST_Request $request Request data.
+	 * @return \WP_REST_Response|WP_Error
+	 */
+	public function get_items( $request ) {
+		$query_args = $this->prepare_reports_query( $request );
+		$query      = $this->construct_query( $query_args );
+		try {
+			$report_data = $query->get_data();
+		} catch ( ParameterException $e ) {
+			return new WP_Error( $e->getErrorCode(), $e->getMessage(), array( 'status' => $e->getCode() ) );
+		}
+
+		$out_data = array(
+			'totals'    => $report_data->totals ? get_object_vars( $report_data->totals ) : null,
+			'intervals' => array(),
+		);
+
+		foreach ( $report_data->intervals as $interval_data ) {
+			$item                    = $this->prepare_item_for_response( $interval_data, $request );
+			$out_data['intervals'][] = $this->prepare_response_for_collection( $item );
+		}
+
+		return $this->add_pagination_headers(
+			$request,
+			$out_data,
+			(int) $report_data->total,
+			(int) $report_data->page_no,
+			(int) $report_data->pages
+		);
+	}
+
+	/**
+	 * Maps query arguments from the REST request, to be fed to Query.
+	 *
+	 * `WP_REST_Request` does not expose a method to return all params covering defaults,
+	 * as it does for `$request['param']` accessor.
+	 * Therefore, we re-implement defaults resolution.
+	 *
+	 * @param \WP_REST_Request $request Full request object.
+	 * @return array Simplified array of params.
+	 */
+	protected function prepare_reports_query( $request ) {
+		$args = wp_parse_args(
+			array_intersect_key(
+				$request->get_query_params(),
+				$this->get_collection_params()
+			),
+			$request->get_default_params()
+		);
+
+		return $args;
+	}
```

Comment:
I haven't started the review but just found something while I was working on an example plugin for the integration guide using this branch. I feel like this method should be defined in `GenericController`, or define a placeholder method for it and throws "not implemented" error. What do you think?

Reply:
I lean towards moving this as is for `GenericConmponent`. This way, new Controllers would not have to re-implement it if they are fine with mapping them from `get_collection_params`.

My only doubt is that this change would affect API of some Controllers that implements their own `get_items` and  doesn't use or have `prepare_reports_query` method:
- `plugins/woocommerce/src/Admin/API/Reports/Controller.php`
- `plugins/woocommerce/src/Admin/API/Reports/Export/Controller.php`
- `plugins/woocommerce/src/Admin/API/Reports/Import/Controller.php`

And due to out long-term backward compatibility policy, I'd be cautious not to add too much.

WDYT?

Reply:
Sorry if I was being unclear, I guess the "throwing not implemented error" idea caused confusion and it isn't the best approach for backward compatibility.

I meant that currently, the prepare_reports_query() method is added to the `GenericStatsController` rather than the `GenericController`. In my case, I created a controller extending `GenericController` without implementing my own `get_items()` method. When I made a request to my controller, it caused a fatal error: `Call to undefined method prepare_reports_query()`.

Since most of the current controllers under `src/Admin/API/Reports/*`, including both stats and non-stats, do not implement their own `get_items()` method, and that the `GenericStatsController` also extends `GenericController`, I think maybe we could move the `prepare_reports_query()` method from `GenericStatsController` to `GenericController`. I might be missing something as I am not fully familiar with this code base, which has many dependencies.

Reply:
Thanks for the clarification. Yes, now I can see the point. Having a Controller that inherits from `GenericController` on the premise of using defaults and throws errors out of the blue is definitely not the experience we want. In such case having a useless method in those 3 classes is acceptable price to pay.

Fixed in https://github.com/woocommerce/woocommerce/pull/49427/commits/c0edda1964ab85f73c6a426e7be90024a11eb14a